benchmarks/auth/python/server.py

```diff
@@ -0,0 +1,127 @@
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from concurrent import futures
+import logging
+
+import grpc
+
+import random
+import string
+
+import os
+import sys
+
+# adding python tracing sources to the system path
+sys.path.insert(0, os.getcwd() + '/../../../../utils/tracing/python')
+import tracing
+import argparse
+
+
+# sys.path.append('../proto')  # for local testing (i.e. not running in Docker-compose)
+import auth_pb2
+import auth_pb2_grpc
+
+import ctypes
+libc = ctypes.CDLL(None)
+syscall = libc.syscall
+
+print("python version: %s" % sys.version)
+print("Server has PID: %d" % os.getpid())
+GRPC_PORT_ADDRESS = os.getenv("GRPC_PORT")
+
+if tracing.IsTracingEnabled():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-a", "--addr", dest="addr", default="0.0.0.0", help="IP address")
+    parser.add_argument("-p", "--port", dest="port", default="50051", help="serve port")
+    parser.add_argument("-zipkin", "--zipkin", dest="url", default="http://0.0.0.0:9411/api/v2/spans", help="Zipkin endpoint url")
+
+    args = parser.parse_args()
+    tracing.initTracer("auth", url=args.url)
+    tracing.grpcInstrumentClient()
+    tracing.grpcInstrumentServer()
+
+class Empty:  # declare this class to enable dynamically adding new attributes
+    pass
+
+def generatePolicy(principalId, effect, resource):
+    authResponse = Empty()
+
+    authResponse.principalId = principalId
+    if effect and resource:
+        policyDocument = Empty()
+        policyDocument.Version = '2012-10-17'
+        policyDocument.Statement = [None]
+        statementOne = Empty()
+        statementOne.Action = 'execute-api:Invoke'
+        statementOne.Effect = effect
+        statementOne.Resource = resource
+        policyDocument.Statement[0] = statementOne
+        authResponse.policyDocument = policyDocument
+    
+    # Optional output with custom properties of the String, Number or Boolean type.
+    authResponse.context = {
+      "stringKey": "stringval",
+      "numberKey": 123,
+      "booleanKey": True
+    }
+
+    return authResponse
+
+
+
+
+class Greeter(auth_pb2_grpc.GreeterServicer):
+
+    def SayHello(self, request, context):
+
+        token = request.name
+        fakeMethodArn = "arn:aws:execute-api:{regionId}:{accountId}:{apiId}/{stage}/{httpVerb}/[{resource}/[{child-resources}]]"
+        
+        with tracing.Span("Generate Policy"):
```

Comment:
Please implement the example functionality from AWS.
So switch on `allow` and `deny` tokens instead of `.f1` and then we should return the generated policy as string if possible. Can you check if that works?

```
// A simple token-based authorizer example to demonstrate how to use an authorization token 
// to allow or deny a request. In this example, the caller named 'user' is allowed to invoke 
// a request if the client-supplied token value is 'allow'. The caller is not allowed to invoke 
// the request if the token value is 'deny'. If the token value is 'unauthorized' or an empty
// string, the authorizer function returns an HTTP 401 status code. For any other token value, 
// the authorizer returns an HTTP 500 status code. 
// Note that token values are case-sensitive.

exports.handler =  function(event, context, callback) {
    var token = event.authorizationToken;
    switch (token) {
        case 'allow':
            callback(null, generatePolicy('user', 'Allow', event.methodArn));
            break;
        case 'deny':
            callback(null, generatePolicy('user', 'Deny', event.methodArn));
            break;
        case 'unauthorized':
            callback("Unauthorized");   // Return a 401 Unauthorized response
            break;
        default:
            callback("Error: Invalid token"); // Return a 500 Invalid token response
    }
};
```